---
title: "Adv_Bioinfo - Challenge 5"
output: html_document
author: Shailesh Dudala
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First create a project directory with 3 directories and make 3 different scripts for the 3 different runs with different resources.
Then copy the fasta files to your project directory.
```{bash, eval=FALSE}
mkdir R; cd R; mkdir R1 R2 R3
cp /gpfs/data/mscbmi/hw4/* .
```
R1 : 1 nodes; 4 cpus; 4gb of RAM
R2 : 1 nodes; 8 cpus; 4gb of RAM
R3 : 1 nodes; 16 cpus; 8gb of RAM

##### B. Develop an analysis pipeline that uses CRI's GARDNER cluster to run the code
```{bash, eval =FALSE}
#For R1: (submitR1.pbs)
[t.cri.biowksp13@cri16in001 R]$ cat submitR1.pbs 
#!/bin/bash

#PBS -N rnaseq_tutorial
#PBS -l mem=4gb,nodes=1:ppn=4,walltime=1:00:00
#PBS -e R1/error-log-${PBS_JOBID}.txt
#PBS -o R1/output-log-${PBS_JOBID}.txt


module load gcc
module load R

Rscript --vanilla ~/R/Rsrc.R
```

```{bash, eval=FALSE}
#For R2: (submitR2.pbs)
[t.cri.biowksp13@cri16in001 R]$ cat submitR2.pbs 
#!/bin/bash

#PBS -N rnaseq_tutorial
#PBS -l mem=4gb,nodes=1:ppn=8,walltime=1:00:00
#PBS -e R2/error-log-${PBS_JOBID}.txt
#PBS -o R2/output-log-${PBS_JOBID}.txt


module load gcc
module load R

Rscript --vanilla ~/R/Rsrc.R

```

```{bash, eval=FALSE}
#For R3: (submitR3.pbs)
[t.cri.biowksp13@cri16in001 R]$ cat submitR3.pbs 
#!/bin/bash

#PBS -N rnaseq_tutorial
#PBS -l mem=16gb,nodes=1:ppn=8,walltime=1:00:00
#PBS -e R3/error-log-${PBS_JOBID}.txt
#PBS -o R2/output-log-${PBS_JOBID}.txt


module load gcc
module load R

Rscript --vanilla ~/R/Rsrc.R
```

##### C. Sort the 'score' value for each aligned sequence and save the sorted scores in a file named: 'scores.txt'
The R script is saved as Rsrc.R
Now, let's sort the scores.
```{bash, eval=FALSE}
grep "score" alignment.txt | sort > scores.txt
```

##### D. What is the maximun and minimum alignment scores for reference.fasta?
```{bash, eval=FALSE}
#Highest
[t.cri.biowksp13@cri16in001 R1]$ tail -1 scores.txt
score: 997.996
```
 

```{bash, eval=FALSE}
#Lowest
[t.cri.biowksp13@cri16in001 R1]$ head -1 scores.txt
score: -0.4660339


```


##### E. Record, report and explain the total runtime (execution time) for your script running with the following hardware configurations: 
```{bash, eval=FALSE}
#R1 : 1 nodes; 4 cpus; 4gb of RAM
[t.cri.biowksp13@cri16in001 R1]$ cat timestamp.txt 
"x"
"1" 3.24429035981496
```

```{bash, eval=FALSE}
#R2: 1 nodes; 8 cpus; 4gb of RAM
[t.cri.biowksp13@cri16in001 R1]$ cd ../R2; cat timestamp.txt
"x"
"1" 3.41031207243601
```

```{bash, eval=FALSE}
#R3: 1 nodes; 16 cpus; 8gb of RAM
[t.cri.biowksp13@cri16in001 R2]$ cd ../R3; cat timestamp.txt
"x"
"1" 3.25919539928436
```

##### F. Are there opportunities for reducing the runtime using Parallelization? If so, what could be the strategy? Algorithm?

>While the observed run times with different resources for computation doesn't seem to be significantly different from each other. Therefore, parallization of the Rscript itself might not be associated with an improvement in efficieny through parallization.

>However, there is an opportunity to parallelize or hyperthread the for loop within the Rscript using the "foreach" and "doParallel" packages, although there isn't necessarily and need for that approach as the function inside the loop isn't as computationally intensive to affect run-time.

>Finally, I would like to add that if the complete pipeline is attempted to be parallelized, then the scores.txt and highest/lowest scores can be a seperate job that wait for alignment.txt to be created first before they execute, illustrated as follows.


```{bash, eval=FALSE}
#First create 2 different jobs with one as the R script (Rsrc.pbs) and the other one to calulate the highest and lowest score (scoresHigh_Low.pbs)

#!/bin/bash
#submit job1 with no dependencies
job1=$(qsub Rsrc.pbs)
echo $job1

#submit job 2 only after job1 has finished
job2=$(qsub -W depend=afterok:$job1 scoresHigh_Low.pbs)
echo $job2


```

